src/data-analysis-environments.rst

```diff
@@ -0,0 +1,322 @@
+    :filter: key == 'kandel:2011'
+
+Using PBD for Data Cleaning
+---------------------------
+Despite the benefits of suggesting transformations in Wrangler, data science workers are limited to preset rules.
+PBD can be used for generating more intelligent suggestions.
+**FlashFill** is a classical example of using programming synthesis to accelerate the data cleaning process :cite:`gulwani:2011`.
+FlashFill takes input-output examples from users and synthesizes a program in a string expression language based on the algorithm.
+The algorithm is able to produce accurate and efficient results.
+It also provides noise detection and an interactive model where users are prompted to provide outputs for ambiguous inputs.
+
+.. raw:: html
+
+    <div style="position: relative; height: 0; overflow: hidden; max-width: 100%; height: auto;">
+    <iframe width="678" height="381" src="https://www.youtube.com/embed/ulbalvFcAYk" frameborder="0" allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture" allowfullscreen></iframe>
+    </div>
+
+.. container:: bib-item
+
+  .. bibliography:: data-analysis-environments.bib
+    :filter: key == 'gulwani:2011'
+
+Using Crowdsourcing for Data Cleaning
+-------------------------------------
+PBD can address some of the cleaning tasks, but not all.
+For example, it is hard to use an algorithmic approach to perfectly solve the problem of entity resolution.
+Entity resolution refers to merging and combining different representations for the same real-world entity.
+For example, a dataset may contain values for "University of Michigan", "U of M", "Umich", "UM".
+They identically refer to one entity and data science workers need to replace all the ambiguous values with the same one.
+It can be difficult to solve by an algorithm but can be easily judged by human knowledge.
+A popular approach for these tasks is crowdsourcing - asking human workers to combine the entities.
+However, this may not be scalable. 
+When the size of the dataset is large, it becomes difficult for human workers to split and coordinate the work.
+
+CrowdER addresses this with a hybrid human-machine approach :cite:`wang:2012`.
+On a high level, it first uses an algorithm to filter and cluster potential values with high probabilities of duplicated entities that need crowd workers to judge.
+Instead of showing crowd works pairwise comparisons, it shows cluster-based values and asks crowd workers to group these candidates.
+This hybrid approach improves both efficiency and accuracy for the result.
+
+.. container:: bib-item
+
+  .. bibliography:: data-analysis-environments.bib
+    :filter: key == 'wang:2012'
+
+Tools for Engineering Features
+==============================
+Finally, the core stage of data science is engineering features, which includes understanding patterns and relations in the data, exploring different representations of the data, and finally extracting insights from the data.
```

Comment:
"Feature engineering" seems to be an odd name for all of these activities -- my understanding was that feature engineering was more about finding features that are used for subsequent data analysis activities. is this definition of "Feature engineering" from somewhere in particular?